src/query_stage.rs

```diff
@@ -99,14 +100,23 @@ impl QueryStage {
     /// Get the input partition count. This is the same as the number of concurrent tasks
     /// when we schedule this query stage for execution
     pub fn get_input_partition_count(&self) -> usize {
-        if self.plan.children().is_empty() {
-            // leaf node (file scan)
-            self.plan.output_partitioning().partition_count()
-        } else {
-            self.plan.children()[0]
-                .output_partitioning()
-                .partition_count()
+        let mut output_partition_counts = HashSet::new();
```

Comment:
We can do this check without creating additional data structures:

        let partition_count = self.plan.children()[0].output_partitioning().partition_count();
        for i in 1..self.plan.children().len() {
            if self.plan.children()[i].output_partitioning().partition_count() != partition_count {
                return Err(...)
            }
        }

Reply:
Makes sense, do you think we should r change the signature to return an error ? That was my original idea, I think there is a 'plan_err!' macro in datafusion we can use

Reply:
PR https://github.com/apache/datafusion-ray/pull/54 no longer looks at the plan's children in this section of code, so perhaps this check (which is a good check) should move somewhere else. I'd suggest adding it in `QueryStage::new` and have it return `Result<Self>`.

Reply:
Makes sense, so let's postpone this as an improvement on #54

Reply:
I converted it into Draft in the meantime